### PR TITLE
fix search fields parameter typing and behavior

### PR DESCRIPTION
**Pull request by philvarner**
**Related Issue(s):** #171 

**Description:**

- fix `fields` typing
- fix `fields` behavior for GET and POST

**PR Checklist:**

- [X] Code is formatted
- [X] Tests pass
- [X] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac-api-client/blob/main/CHANGELOG.md)

### CODE REVIEW

File: pystac_client/item_search.py [hunk 1]
```diff
@@ -431,9 +434,32 @@ def _format_fields(self, value: Optional[FieldsLike]) -> Optional[Fields]:
...
         if isinstance(value, str):
-            return tuple(value.split(','))
+            return self.fields_to_json(value.split(','))
+        if isinstance(value, list):
+            return self.fields_to_json(value)
+        if isinstance(value, dict):
+            return value
 
-        return tuple(value)
+        raise Exception("sortby must be of type None, str, List[str], or List[Dict[str, str]")
```
Comment (gadomski) [inline pystac_client/item_search.py:443]:
```suggestion
        raise Exception("fields must be of type None, str, List[str], or List[Dict[str, str]")
```

File: pystac_client/item_search.py [hunk 2]
```diff
@@ -431,9 +434,32 @@ def _format_fields(self, value: Optional[FieldsLike]) -> Optional[Fields]:
...
+            if field.startswith("-"):
+                excludes.append(field[1:])
+            elif field.startswith("+"):
+                includes.append(field[1:])
+            else:
+                includes.append(field)
+        return {"includes": includes, "excludes": excludes}
+
+    @staticmethod
+    def fields_json_to_str(fields: Fields) -> str:
```
Comment (gadomski) [inline pystac_client/item_search.py:459]: Same as above, it's not really JSON, it's a dict.

File: pystac_client/item_search.py [hunk 3]
```diff
@@ -54,8 +55,8 @@
 Sortby = List[Dict[str, str]]
 SortbyLike = Union[Sortby, str, List[str]]
 
-Fields = List[str]
-FieldsLike = Union[Fields, str]
+Fields = Dict[str, List[str]]
```
Comment (gadomski) [inline pystac_client/item_search.py:58]: There's enough `staticmethod` logic around `Fields` that you _could_ make the case for `Fields` to be its own class, with a `parse` classmethod and a `to_string` method. I don't think it's required, just a thought.

**Review by gadomski — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +17,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Fixed
 
 - Search sortby parameter now has correct typing and correctly handles both GET and POST JSON parameter formats. [#175](https://github.com/stac-utils/pystac-client/pull/175)
+- Search fields parameter now has correct typing and correctly handles both GET and POST JSON parameter formats.
```
[the remark this answers is not in the document]
Comment (philvarner) [reply CHANGELOG.md:20]: no, one's for sortby and one for fields

[the remark this answers is not in the document]
Comment (gadomski) [reply CHANGELOG.md:20]: Ah gotcha. Well, missing pull request link then :-P

File: pystac_client/item_search.py (hunk 3 above)
Comment (philvarner) [reply pystac_client/item_search.py:65]: I agree -- I think there's a good deal of refactoring that could be done here with validation. For now, I just wanted to get the existing patterns working correctly

File: pystac_client/item_search.py [hunk 5]
```diff
@@ -431,9 +434,32 @@ def _format_fields(self, value: Optional[FieldsLike]) -> Optional[Fields]:
...
+        if isinstance(value, list):
+            return self.fields_to_json(value)
+        if isinstance(value, dict):
+            return value
 
-        return tuple(value)
+        raise Exception("sortby must be of type None, str, List[str], or List[Dict[str, str]")
+
+    @staticmethod
+    def fields_to_json(fields: List[str]) -> Fields:
```
[the remark this answers is not in the document]
Comment (philvarner) [reply pystac_client/item_search.py]: good point.

**Pull request merged**